src/aktualizr_secondary/aktualizr_secondary.cc

```diff
@@ -128,23 +130,46 @@ bool AktualizrSecondary::sendFirmwareResp(const std::shared_ptr<std::string>& fi
 #endif
   } else if (pacman->name() == "debian") {
     // TODO save debian package here.
-    LOG_ERROR << "Installation of non ostree images is not suppotrted yet.";
+    LOG_ERROR << "Installation of debian images is not suppotrted yet.";
     return false;
   }
 
-  // This check is not applicable for the ostree case, just for the binary file case(s)
-  if (targetToApply.length() != firmware->length()) {
+  if (targetToApply.length() != firmware_size) {
     LOG_ERROR << "The target image size specified in metadata " << targetToApply.length()
               << " does not match actual size " << firmware->length();
     return false;
   }
 
+  if (!targetToApply.IsOstree()) {
+    Uptane::Hash sha256_hash(Uptane::Hash::Type::kSha256,
+                             boost::algorithm::to_lower_copy(boost::algorithm::hex(Crypto::sha256digest(*firmware))));
+    Uptane::Hash sha512_hash(Uptane::Hash::Type::kSha512,
+                             boost::algorithm::to_lower_copy(boost::algorithm::hex(Crypto::sha512digest(*firmware))));
+
+    if (!(targetToApply.MatchHash(sha256_hash) && targetToApply.MatchHash(sha512_hash))) {
+      LOG_ERROR << "The received image data hash doesn't match the hash specified in the target metadata";
+      return false;
+    }
```

Comment:
I think it's worth reporting which hash has failed in the error message. Will make code slightly bigger but will save headaches later on.

Reply:
Realistically, they will both match or not. Logging the third imaginary case does not sound so important.

Reply:
I disagree: I think this case sould be explicitly tested for in tuf-test-vectors (at least) and hence logging will come in handy.

Reply:
No problem, will add more details into the log message.